File: app/src/main/java/it/davidenastri/littlecloud/QueryUtils.java
package it.davidenastri.littlecloud;

import android.content.Context;
import android.content.SharedPreferences;
import android.util.Log;
import android.view.View;

import com.google.android.material.snackbar.Snackbar;
import com.loopj.android.http.AsyncHttpClient;
import com.loopj.android.http.RequestParams;
import com.loopj.android.http.TextHttpResponseHandler;

import cz.msebera.android.httpclient.Header;

public final class QueryUtils {

    private static final String SETTINGS = "SETTINGS";
    private static final String PARTICLE_DEVICE_ID_KEY = "particle_device_id";
    private static final String PARTICLE_TOKEN_ID_KEY = "particle_token_id";
    private static final String PARTICLE_API_URL_KEY = "particle_token_url";
    private static final String DEFAULT_NAME = "defaultName";
    private static final String LOG_TAG = "QueryUtils";

    private QueryUtils() {
    }

    private static SharedPreferences getSharedPreferences(View view) {
        return view.getContext().getSharedPreferences(SETTINGS, Context.MODE_PRIVATE);
    }

    private static String[] getParticleDetails(View view) {
        SharedPreferences sharedPreferences = getSharedPreferences(view);
        String deviceId = sharedPreferences.getString(PARTICLE_DEVICE_ID_KEY, DEFAULT_NAME);
        String tokenId = sharedPreferences.getString(PARTICLE_TOKEN_ID_KEY, DEFAULT_NAME);
        String apiUrl = sharedPreferences.getString(PARTICLE_API_URL_KEY, DEFAULT_NAME);
        return new String[]{deviceId, tokenId, apiUrl};
    }

    private static RequestParams createRequestParams(String tokenId, String args) {
        RequestParams params = new RequestParams();
        params.put("access_token", tokenId);
        params.put("args", args);
        return params;
    }

    public static void changeColor(String rgbString, final String colorSet, final View onClickView) {
        String[] particleDetails = getParticleDetails(onClickView);
        String deviceId = particleDetails[0];
        String tokenId = particleDetails[1];
        String apiUrl = particleDetails[2];
        RequestParams params = createRequestParams(tokenId, rgbString);
        AsyncHttpClient client = new AsyncHttpClient();

        client.post(apiUrl + deviceId + "/setColor", params, new TextHttpResponseHandler() {
            @Override
            public void onSuccess(int statusCode, Header[] headers, String res) {
                Log.i(LOG_TAG, "Color set to: " + colorSet);
                Snackbar.make(onClickView, "Lamp color changed successfully!", Snackbar.LENGTH_LONG).show();
            }

            @Override
            public void onFailure(int statusCode, Header[] headers, String res, Throwable t) {
                Log.e(LOG_TAG, "Failed to set color: " + res);
                Snackbar.make(onClickView, "Little Cloud is offline...", Snackbar.LENGTH_LONG).show();
            }
        });
    }

    public static void changeAudio(final String commandString, final View onClickView) {
        String[] particleDetails = getParticleDetails(onClickView);
        String deviceId = particleDetails[0];
        String tokenId = particleDetails[1];
        String apiUrl = particleDetails[2];
        RequestParams params = createRequestParams(tokenId, commandString);
        AsyncHttpClient client = new AsyncHttpClient();

        client.post(apiUrl + deviceId + "/dfMini", params, new TextHttpResponseHandler() {
            @Override
            public void onSuccess(int statusCode, Header[] headers, String res) {
                Log.i(LOG_TAG, "Audio command sent: " + commandString);
                Snackbar.make(onClickView, "Lamp audio changed successfully!", Snackbar.LENGTH_LONG).show();
            }

            @Override
            public void onFailure(int statusCode, Header[] headers, String res, Throwable t) {
                Log.e(LOG_TAG, "Failed to change audio: " + res);
                Snackbar.make(onClickView, "Little Cloud is offline...", Snackbar.LENGTH_LONG).show();
            }
        });
    }
}

File: app/src/main/java/it/davidenastri/littlecloud/QueryUtils.kt
package it.davidenastri.littlecloud

import android.content.Context
import android.content.SharedPreferences
import android.util.Log
import android.view.View
import android.widget.Toast
import it.davidenastri.littlecloud.network.ParticleApiService
import it.davidenastri.littlecloud.network.RetrofitClient
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

object QueryUtils {

    private const val SETTINGS = "SETTINGS"
    private const val PARTICLE_DEVICE_ID_KEY = "particleDeviceId"
    private const val PARTICLE_TOKEN_ID_KEY = "particleTokenId"
    private const val PARTICLE_API_URL_KEY = "particleApiUrl"
    private const val DEFAULT_NAME = "defaultName"
    private const val LOG_TAG = "QueryUtils"

    @Volatile
    private var isRequestInProgress = false

    @Synchronized
    private fun startRequest(): Boolean {
        if (isRequestInProgress) return false
        isRequestInProgress = true
        return true
    }

    @Synchronized
    private fun endRequest() {
        isRequestInProgress = false
    }

    private fun getSharedPreferences(view: View): SharedPreferences {
        Log.d(LOG_TAG, "Getting SharedPreferences for view: $view")
        val sharedPreferences = view.context.getSharedPreferences(SETTINGS, Context.MODE_PRIVATE)
        Log.d(LOG_TAG, "SharedPreferences content: ${sharedPreferences.all}")
        return sharedPreferences
    }

    private fun getParticleDetails(view: View): Array<String> {
        Log.d(LOG_TAG, "Fetching particle details from SharedPreferences")
        val sharedPreferences = getSharedPreferences(view)
        val deviceId = sharedPreferences.getString(PARTICLE_DEVICE_ID_KEY, DEFAULT_NAME) ?: DEFAULT_NAME
        val tokenId = sharedPreferences.getString(PARTICLE_TOKEN_ID_KEY, DEFAULT_NAME) ?: DEFAULT_NAME
        val apiUrl = sharedPreferences.getString(PARTICLE_API_URL_KEY, DEFAULT_NAME) ?: DEFAULT_NAME
        Log.d(LOG_TAG, "Particle details - Device ID: $deviceId, Token ID: $tokenId, API URL: $apiUrl")
        return arrayOf(deviceId, tokenId, apiUrl)
    }

    private fun getErrorMessage(statusCode: Int): String {
        return when (statusCode) {
            0 -> "No network connection. Please check your internet."
            401 -> "Invalid access token. Please check settings."
            404 -> "Device not found. Please check Device ID."
            408 -> "Request timeout. Please try again."
            in 500..599 -> "Server error. Please try again later."
            else -> "Failed to communicate with lamp. Error: $statusCode"
        }
    }

    fun changeColor(rgbString: String, colorSet: String, onClickView: View) {
        if (!startRequest()) {
            Toast.makeText(onClickView.context, "Please wait, request is already in progress.", Toast.LENGTH_SHORT).show()
            return
        }

        Log.d(LOG_TAG, "Changing color to: $colorSet with RGB string: $rgbString")
        val (deviceId, tokenId, apiUrl) = getParticleDetails(onClickView)

        val apiService = RetrofitClient.getClient(apiUrl).create(ParticleApiService::class.java)

        CoroutineScope(Dispatchers.IO).launch {
            try {
                val response = apiService.setColor(deviceId, tokenId, rgbString)
                withContext(Dispatchers.Main) {
                    endRequest()
                    Log.i(LOG_TAG, "Color set to: $colorSet successfully. Response: $response")
                    val message = if (response.connected && response.return_value == 1) {
                        "Lamp color changed successfully! Lamp is online."
                    } else {
                        "Lamp color changed, but the lamp might be offline."
                    }
                    Toast.makeText(onClickView.context, message, Toast.LENGTH_LONG).show()
                }
            } catch (e: Exception) {
                withContext(Dispatchers.Main) {
                    endRequest()
                    Log.e(LOG_TAG, "Failed to set color", e)
                    Toast.makeText(onClickView.context, "Failed to communicate with lamp.", Toast.LENGTH_LONG).show()
                }
            }
        }
    }

    fun changeAudio(commandString: String, onClickView: View) {
        if (!startRequest()) {
            Toast.makeText(onClickView.context, "Please wait, request is already in progress.", Toast.LENGTH_SHORT).show()
            return
        }

        Log.d(LOG_TAG, "Changing audio with command: $commandString")
        val (deviceId, tokenId, apiUrl) = getParticleDetails(onClickView)

        val apiService = RetrofitClient.getClient(apiUrl).create(ParticleApiService::class.java)

        CoroutineScope(Dispatchers.IO).launch {
            try {
                val response = apiService.controlAudio(deviceId, tokenId, commandString)
                withContext(Dispatchers.Main) {
                    endRequest()
                    Log.i(LOG_TAG, "Audio command sent: $commandString successfully. Response: $response")
                    val message = if (response.connected && response.return_value == 1) {
                        "Lamp audio changed successfully! Lamp is online."
                    } else {
                        "Lamp audio changed, but the lamp might be offline."
                    }
                    Toast.makeText(onClickView.context, message, Toast.LENGTH_LONG).show()
                }
            } catch (e: Exception) {
                withContext(Dispatchers.Main) {
                    endRequest()
                    Log.e(LOG_TAG, "Failed to change audio", e)
                    Toast.makeText(onClickView.context, "Failed to communicate with lamp.", Toast.LENGTH_LONG).show()
                }
            }
        }
    }
}
